src/Aspire.Hosting/Extensions/ResourceBuilderExtensions.cs

```diff
@@ -359,9 +360,23 @@ private static void ApplyEndpoints<T>(this IResourceBuilder<T> builder, IResourc
     /// <param name="endpointName">Name of endpoint to change.</param>
     /// <param name="callback">Callback that modifies the endpoint.</param>
     /// <param name="createIfNotExists">Create endpoint if it does not exist.</param>
+    /// <param name="deferred">Defers execution of callback until before start.</param>
     /// <returns></returns>
-    public static IResourceBuilder<T> WithEndpoint<T>(this IResourceBuilder<T> builder, string endpointName, Action<EndpointAnnotation> callback, bool createIfNotExists = true) where T : IResourceWithEndpoints
+    public static IResourceBuilder<T> WithEndpoint<T>(this IResourceBuilder<T> builder, string endpointName, Action<EndpointAnnotation> callback, bool createIfNotExists = true, bool deferred = false) where T : IResourceWithEndpoints
     {
```

Comment:
Looks like it's separate on purpose https://github.com/dotnet/aspire/issues/2169#issue-2128077310
Although I don't know if there is any specific scenario for the immediate one?

I guess if you want to add an endpoint to a resource with the callback then you need to not defer it. Or we need to change this PR to create an endpoint annotation for any callbacks that weren't called for existing endpoints.

Reply:
This change was made to be fairly conservative. I hadn't fully thought through whether changing everything to deferred would have unintended consequences. For example in some places we use endpoint annotations to build config files which are then mounted to containers. If the endpoint mutation doesn't run before that then things will be missed.

So there are definately timing issues here if we make everything deferred. There are also some differences in where endpoints are injected between run mode vs. publish mode which this PR doesn't take into account (yet).

Reply:
I would need some pointers for where all the timing issues might be. It seems like right now the callbacks run pretty early, but I'm not fully aware of everything that happens in the app host.

Issues I see with always deferring:
1. What do we do with endpoints that don't exist when the callbacks are iterated over?
   - We can create a new `EndpointAnnotation` and pass it to the callback then add it to the resource
2. What if the user just wants to modify an endpoint and not create one?

Issues with having an option to defer or not:
1. Callsite isn't very pretty
2. Still need to handle "endpoint doesn't exist when callbacks iterated over"

Reply:
https://github.com/dotnet/aspire/pull/2448/commits/d7a7e6b7e549d215cd6b5b3fde4b6910e5bf1d19 shows what having no bool options could look like.

Reply:
Problem with creating a new endpoint is you have no idea if you need to set more values or not:
```
builder.AddProject<Projects.MyFrontend>("frontend")
       .WithReference(basketService)
       .WithReference(catalogService.GetEndpoint("http"))
       .WithEndpoint("https", endpoint =>
       {
           endpoint.IsExternal = true;
       });
```
This has errors in the console because "https" doesn't exist by default for frontend, so we create a new endpoint and just set IsExternal, meaning it has a tcp endpoint without a port.